### PR TITLE
[v2] Use an apply configuration for the workspace template

### PR DESCRIPTION
**Pull request by blampe**
This adds `applyconfiguration-gen` to generate apply configurations for our APIs. The workspace template is updated to use it.

It's a moderately uphill battle to get apply configurations working with controllers due to various pieces not including built-in support for them. In practice this means we need to cut a couple small corners -- in particular implementing our own `DeepCopy` and tweaking the generated `GroupVersion` code.

Fixes https://github.com/pulumi/pulumi-kubernetes-operator/issues/704.

### CODE REVIEW

File: operator/internal/controller/pulumi/session_test.go [hunk 1]
```diff
@@ -357,17 +359,17 @@ func TestSetupWorkspace(t *testing.T) {
...
-					Spec: &autov1alpha1.WorkspaceSpec{
-						Image:              "custom-image",
-						ServiceAccountName: "custom-service-account",
-						PodTemplate: &autov1alpha1.EmbeddedPodTemplateSpec{
-							Metadata: autov1alpha1.EmbeddedObjectMeta{
+					Spec: &apply.WorkspaceSpecApplyConfiguration{
+						Image:              ptr.To("custom-image"),
+						ServiceAccountName: ptr.To("custom-service-account"),
+						PodTemplate: &apply.EmbeddedPodTemplateSpecApplyConfiguration{
+							Metadata: &apply.EmbeddedObjectMetaApplyConfiguration{
```
Comment (blampe) [inline operator/internal/controller/pulumi/session_test.go:372]: If we're happy with the direction of this I can take it further and also use an apply config for our pod template -- it should be very minor.

File: operator/internal/controller/pulumi/session_test.go [hunk 2]
```diff
@@ -357,17 +359,17 @@ func TestSetupWorkspace(t *testing.T) {
 						},
 					},
 				},
-				WorkspaceTemplate: &autov1alpha1.EmbeddedWorkspaceTemplateSpec{
-					Metadata: autov1alpha1.EmbeddedObjectMeta{
+				WorkspaceTemplate: &shared.WorkspaceApplyConfiguration{
+					ObjectMetaApplyConfiguration: &metav1apply.ObjectMetaApplyConfiguration{
```
Comment (blampe) [inline operator/internal/controller/pulumi/session_test.go:363]: Worth calling out that `ObjectMetaApplyConfiguration` differs from `EmbeddedWorkspaceTemplateSpec` in that it does expose all of the underlying fields (like generation) even if they can't actually be set. I think it makes for a more bloated/confusing schema but I don't see a good way around it.

File: operator/internal/controller/pulumi/session_test.go (hunk 1 above)
Comment (EronWright) [reply operator/internal/controller/pulumi/session_test.go:372]: That would be pretty wonderful.

**Review by EronWright — APPROVED** [review]
Awesome!  Good job finding a workaround for the DeepCopy issue.

File: operator/internal/controller/pulumi/session_test.go (hunk 2 above)
Comment (rquitales) [reply operator/internal/controller/pulumi/session_test.go:363]: This approach is similar to `PodTemplateSpec` so it shouldn't add confusion to K8s users.

**Review by rquitales — APPROVED** [review]

File: operator/internal/controller/pulumi/session_test.go (hunk 1 above)
Comment (blampe) [reply operator/internal/controller/pulumi/session_test.go:372]: I spoke too soon -- this is actually quite involved. I'll put up a WIP with what I have so far.

**Pull request merged**